iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/thrift/SyncClientManager.java

```diff
@@ -0,0 +1,232 @@
+      String trustStore,
+      String trustStorePwd,
+      boolean useLeaderCache) {
+    this.useSSL = useSSL;
+    this.trustStore = trustStore;
+    this.trustStorePwd = trustStorePwd;
+    this.useLeaderCache = useLeaderCache;
+    this.endPoints = endPoints;
+
+    for (TEndPoint endPoint : endPoints) {
+      endPoint2client.put(endPoint, new Pair<>(null, false));
+    }
+  }
+
+  public void initClients() throws IOException, TTransportException {
+    // init clients
+    for (TEndPoint endPoint : endPoints) {
+      if (Boolean.TRUE.equals(endPoint2client.get(endPoint).getRight())) {
+        continue;
+      }
+
+      initAndSetClient(endPoint);
+    }
+    // check whether any clients are available
+    for (TEndPoint nodeUrl : endPoint2client.keySet()) {
+      if (Boolean.TRUE.equals(endPoint2client.get(nodeUrl).getRight())) {
+        return;
+      }
+    }
+
+    throw new PipeConnectionException(
+        String.format("All target servers %s are not available.", endPoint2client.keySet()));
+  }
+
+  public Pair<IoTDBThriftSyncConnectorClient, Boolean> getOneConnectedClientAndStatus() {
+    final int clientSize = endPoints.size();
+    // Round-robin, find the next alive client
+    for (int tryCount = 0; tryCount < clientSize; ++tryCount) {
+      final int clientIndex = (int) (currentClientIndex++ % clientSize);
+      Pair<IoTDBThriftSyncConnectorClient, Boolean> clientAndStatus =
+          endPoint2client.get(endPoints.get(clientIndex));
+      if (Boolean.TRUE.equals(clientAndStatus.getRight())) {
+        return clientAndStatus;
+      }
+    }
+    throw new PipeConnectionException(
+        "All clients are dead, please check the connection to the receiver.");
+  }
+
+  public Pair<IoTDBThriftSyncConnectorClient, Boolean> getClientAndStatusByEvent(
+      PipeInsertNodeTabletInsertionEvent event) {
+    TEndPoint endPoint = leaderCacheManager.parseEndPoint(event);
+    return endPoint == null
+            || !useLeaderCache
+            || !endPoint2client.containsKey(endPoint)
+            || Boolean.FALSE.equals(endPoint2client.get(endPoint).getRight())
+        ? getOneConnectedClientAndStatus()
+        : endPoint2client.get(endPoint);
+  }
+
+  public Pair<IoTDBThriftSyncConnectorClient, Boolean> getClientAndStatusByEvent(
+      PipeRawTabletInsertionEvent event) {
+    TEndPoint endPoint = leaderCacheManager.parseEndPoint(event);
+    return endPoint == null
+            || !useLeaderCache
+            || !endPoint2client.containsKey(endPoint)
+            || Boolean.FALSE.equals(endPoint2client.get(endPoint).getRight())
+        ? getOneConnectedClientAndStatus()
+        : endPoint2client.get(endPoint);
+  }
+
+  public Pair<IoTDBThriftSyncConnectorClient, Boolean> getClientAndStatusByEvent(
+      PipeTsFileInsertionEvent event) {
+    return getOneConnectedClientAndStatus();
+  }
+
+  public void updateOrCreate(String deviceId, TEndPoint endPoint) {
+    try {
+      endPoints.add(endPoint);
```

Comment:
The "endPoints" may be useless after the first call to "initClients()", by the way, if you "add" an endpoint to update that you may be expecting the "endPoints" to be a set.